# Ex1 - Filtrar y ordenar datos

### Paso 1. Importar las bibliotecas necesarias

In [64]:
import pandas as pd
import numpy as np

### Paso 2. Importar el conjunto de datos de esta [dirección](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv). 

In [65]:
chipo=pd.read_csv(r'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv',delimiter='\t')

### Paso 3. Asígnalo a una variable llamada chipo.

In [66]:
chipo.head(20)

order_id  quantity                              item_name  \
0          1         1           Chips and Fresh Tomato Salsa   
1          1         1                                   Izze   
2          1         1                       Nantucket Nectar   
3          1         1  Chips and Tomatillo-Green Chili Salsa   
4          2         2                           Chicken Bowl   
5          3         1                           Chicken Bowl   
6          3         1                          Side of Chips   
7          4         1                          Steak Burrito   
8          4         1                       Steak Soft Tacos   
9          5         1                          Steak Burrito   
10         5         1                    Chips and Guacamole   
11         6         1                   Chicken Crispy Tacos   
12         6         1                     Chicken Soft Tacos   
13         7         1                           Chicken Bowl   
14         7         1                    Chips and Guacamole   
15         8         1  Chips and Tomatillo-Green Chili Salsa   
16         8         1                        Chicken Burrito   
17         9         1                        Chicken Burrito   
18         9         2                            Canned Soda   
19        10         1                           Chicken Bowl   

                                   choice_description item_price  
0                                                 NaN     $2.39   
1                                        [Clementine]     $3.39   
2                                             [Apple]     $3.39   
3                                                 NaN     $2.39   
4   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                 NaN     $1.69   
7   [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8   [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9   [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25   
10                                                NaN     $4.45   
11  [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75   
12  [Roasted Chili Corn Salsa, [Rice, Black Beans,...     $8.75   
13  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $11.25   
14                                                NaN     $4.45   
15                                                NaN     $2.39   
16  [Tomatillo-Green Chili Salsa (Medium), [Pinto ...     $8.49   
17  [Fresh Tomato Salsa (Mild), [Black Beans, Rice...     $8.49   
18                                           [Sprite]     $2.18   
19  [Tomatillo Red Chili Salsa, [Fajita Vegetables...     $8.75

# Nombre del producto de valor máximo

In [67]:
chipo['item_price'].max()

'$9.39 '

- Parece que mi columna item_price no es del tipo int ya que el valor max no es $9.39 

In [68]:
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


- Observamos que la columna item_price es del tipo object y que no tiene null.

- Miremos que tipos tiene la columna item_price

In [69]:
# la columna no tiene Null 
#chipo[chipo['item_price'].isna()]

chipo['item_price'].apply(type)

0       <class 'str'>
1       <class 'str'>
2       <class 'str'>
3       <class 'str'>
4       <class 'str'>
            ...      
4617    <class 'str'>
4618    <class 'str'>
4619    <class 'str'>
4620    <class 'str'>
4621    <class 'str'>
Name: item_price, Length: 4622, dtype: object

- vemos que la clase es de tipo string 
- vamos a eliminar el simbolo $ u espacios o comas que pueda tener y convertirlo en folat

In [70]:
chipo['item_price'] = chipo['item_price'].replace({'\$': '', ',': ''}, regex=True).astype(float)

In [71]:
# tb se puede hacer 
# chipo['item_price'] =chipo['item_price'].str.replace('$','')
# chipo['item_price'].str.strip().astype(float)

- volvamos a comprobar de que tipo es nuestra columna 

In [72]:
chipo['item_price'].dtype

dtype('float64')

- parece que ahora si es del tipo float 
- vamos a sacar el maximo comprobando que el numero de la columna quantity sea 1 ,
ya que nos interesa que el valor unitario del producto.

In [73]:
chipo[chipo['quantity']==1]['item_price'].max()

np.float64(11.89)

In [74]:
filtrado_1=chipo[chipo['quantity']==1]

precio_max=filtrado_1['item_price'].max()

precio_max

np.float64(11.89)

- como puede haber mas de un producto que sea maximo 
- vamos a ver cuantos hay y quienes son 

In [75]:
precio_max_todos=filtrado_1[filtrado_1['item_price']==precio_max]
precio_max_todos

order_id  quantity            item_name  \
606        250         1     Steak Salad Bowl   
749        309         1     Steak Salad Bowl   
1132       468         1  Carnitas Salad Bowl   
1159       478         1     Steak Salad Bowl   
1229       501         1  Barbacoa Salad Bowl   
1311       534         1     Steak Salad Bowl   
1326       541         1  Barbacoa Salad Bowl   
1505       612         1     Steak Salad Bowl   
1571       637         1     Steak Salad Bowl   
1590       644         1     Steak Salad Bowl   
1816       734         1     Steak Salad Bowl   
2401       953         1     Steak Salad Bowl   
2439       969         1     Steak Salad Bowl   
2442       970         1  Barbacoa Salad Bowl   
2610      1035         1  Carnitas Salad Bowl   
2740      1088         1     Steak Salad Bowl   
2804      1114         1     Steak Salad Bowl   
2957      1176         1     Steak Salad Bowl   
3120      1244         1     Steak Salad Bowl   
3208      1283         1  Barbacoa Salad Bowl   
3350      1343         1     Steak Salad Bowl   
3493      1403         1     Steak Salad Bowl   
3546      1426         1  Barbacoa Salad Bowl   
3749      1499         1  Carnitas Salad Bowl   
4239      1692         1  Carnitas Salad Bowl   
4313      1720         1     Steak Salad Bowl   

                                     choice_description  item_price  
606   [Fresh Tomato Salsa, [Pinto Beans, Cheese, Gua...       11.89  
749   [Roasted Chili Corn Salsa, [Rice, Cheese, Lett...       11.89  
1132  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...       11.89  
1159  [Fresh Tomato Salsa, [Rice, Fajita Vegetables,...       11.89  
1229  [Fresh Tomato Salsa, [Rice, Fajita Vegetables,...       11.89  
1311  [Roasted Chili Corn Salsa, [Fajita Vegetables,...       11.89  
1326  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.89  
1505  [Fresh Tomato Salsa, [Rice, Pinto Beans, Chees...       11.89  
1571  [Fresh Tomato Salsa, [Fajita Vegetables, Sour ...       11.89  
1590  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.89  
1816  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...       11.89  
2401  [Fresh Tomato Salsa, [Fajita Vegetables, Guaca...       11.89  
2439  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.89  
2442  [Tomatillo Green Chili Salsa, [Rice, Fajita Ve...       11.89  
2610  [Roasted Chili Corn Salsa, [Fajita Vegetables,...       11.89  
2740  [Fresh Tomato Salsa, [Fajita Vegetables, Black...       11.89  
2804  [Tomatillo Red Chili Salsa, [Rice, Black Beans...       11.89  
2957  [Fresh Tomato Salsa, [Black Beans, Cheese, Gua...       11.89  
3120  [Roasted Chili Corn Salsa, [Fajita Vegetables,...       11.89  
3208  [Tomatillo Red Chili Salsa, [Black Beans, Chee...       11.89  
3350  [Fresh Tomato Salsa, [Cheese, Guacamole, Lettu...       11.89  
3493  [Roasted Chili Corn Salsa, [Fajita Vegetables,...       11.89  
3546                    [Fresh Tomato Salsa, Guacamole]       11.89  
3749  [Roasted Chili Corn Salsa, [Fajita Vegetables,...       11.89  
4239  [Tomatillo Green Chili Salsa, [Black Beans, Ch...       11.89  
4313  [Roasted Chili Corn Salsa, [Fajita Vegetables,...       11.89

In [76]:
precio_max_todos.shape[0]

26

- observamos que hay 26 variables que tienen el maximo valor 

### Paso 4. ¿Cuántos productos cuestan más de 10,00 dólares?

- interpretamos que nos piden el numero de productos individuales que cuestan mas de 10$ 
- por tanto usaremos el df `filtrado_1` 

In [77]:
mayor_10=filtrado_1[filtrado_1['item_price']>10].nunique()
mayor_10

order_id              753
quantity                1
item_name              25
choice_description    416
item_price              8
dtype: int64

- hay 997 articulos con costo mayor de 10$

### Paso 5. ¿Cuál es el precio de cada artículo?
###### imprimir un marco de datos con sólo dos columnas nombre_artículo y precio_artículo

In [62]:
filtrado_1[['item_name','item_price']]

item_name  item_price
0              Chips and Fresh Tomato Salsa        2.39
1                                      Izze        3.39
2                          Nantucket Nectar        3.39
3     Chips and Tomatillo-Green Chili Salsa        2.39
5                              Chicken Bowl       10.98
...                                     ...         ...
4617                          Steak Burrito       11.75
4618                          Steak Burrito       11.75
4619                     Chicken Salad Bowl       11.25
4620                     Chicken Salad Bowl        8.75
4621                     Chicken Salad Bowl        8.75

[4355 rows x 2 columns]

### Paso 6. Ordenar por el nombre del artículo

- interpretamos que debemos ordenar el df original y no el filtrado 

In [63]:
chipo.sort_values(by='item_name')

order_id  quantity          item_name  \
3389      1360         2  6 Pack Soft Drink   
341        148         1  6 Pack Soft Drink   
1849       749         1  6 Pack Soft Drink   
1860       754         1  6 Pack Soft Drink   
2713      1076         1  6 Pack Soft Drink   
...        ...       ...                ...   
2384       948         1  Veggie Soft Tacos   
781        322         1  Veggie Soft Tacos   
2851      1132         1  Veggie Soft Tacos   
1699       688         1  Veggie Soft Tacos   
1395       567         1  Veggie Soft Tacos   

                                     choice_description  item_price  
3389                                        [Diet Coke]       12.98  
341                                         [Diet Coke]        6.49  
1849                                             [Coke]        6.49  
1860                                        [Diet Coke]        6.49  
2713                                             [Coke]        6.49  
...                                                 ...         ...  
2384  [Roasted Chili Corn Salsa, [Fajita Vegetables,...        8.75  
781   [Fresh Tomato Salsa, [Black Beans, Cheese, Sou...        8.75  
2851  [Roasted Chili Corn Salsa (Medium), [Black Bea...        8.49  
1699  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
1395  [Fresh Tomato Salsa (Mild), [Pinto Beans, Rice...        8.49  

[4622 rows x 5 columns]

### Paso 7. ¿Cuál fue la cantidad pedida del artículo más caro? 2 formas

In [79]:
chipo[chipo['item_price']==chipo['item_price'].max()]['quantity'].sort_values(ascending=False).values[0]

np.int64(15)

In [46]:
precio_max_todos['order_id'].value_counts()

order_id
250     1
309     1
468     1
478     1
501     1
534     1
541     1
612     1
637     1
644     1
734     1
953     1
969     1
970     1
1035    1
1088    1
1114    1
1176    1
1244    1
1283    1
1343    1
1403    1
1426    1
1499    1
1692    1
1720    1
Name: count, dtype: int64

In [23]:
precio_max_todos.groupby('order_id').size()

order_id
250     1
309     1
468     1
478     1
501     1
534     1
541     1
612     1
637     1
644     1
734     1
953     1
969     1
970     1
1035    1
1088    1
1114    1
1176    1
1244    1
1283    1
1343    1
1403    1
1426    1
1499    1
1692    1
1720    1
dtype: int64

### Paso 8. ¿Cuántas veces pidió ensalada de verduras?

In [ ]:
chipo[chipo['item_name']=='Veggie Salad Bowl'].count()

order_id              18
quantity              18
item_name             18
choice_description    18
item_price            18
dtype: int64

In [80]:
chipo[chipo['item_name']=='Veggie Salad Bowl'].count()[0]

C:\Users\denis\AppData\Local\Temp\ipykernel_36576\394205218.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  chipo[chipo['item_name']=='Veggie Salad Bowl'].count()[0]


np.int64(18)

In [55]:
num_pedido=chipo[chipo['item_name']=='Veggie Salad Bowl']
len(num_pedido)

18

In [56]:
num_pedido

order_id  quantity          item_name  \
186         83         1  Veggie Salad Bowl   
295        128         1  Veggie Salad Bowl   
455        195         1  Veggie Salad Bowl   
496        207         1  Veggie Salad Bowl   
960        394         1  Veggie Salad Bowl   
1316       536         1  Veggie Salad Bowl   
1884       760         1  Veggie Salad Bowl   
2156       869         1  Veggie Salad Bowl   
2223       896         1  Veggie Salad Bowl   
2269       913         1  Veggie Salad Bowl   
2683      1066         1  Veggie Salad Bowl   
3223      1289         1  Veggie Salad Bowl   
3293      1321         1  Veggie Salad Bowl   
4109      1646         1  Veggie Salad Bowl   
4201      1677         1  Veggie Salad Bowl   
4261      1700         1  Veggie Salad Bowl   
4541      1805         1  Veggie Salad Bowl   
4573      1818         1  Veggie Salad Bowl   

                                     choice_description  item_price  
186   [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
295   [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...       11.25  
455   [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
496   [Fresh Tomato Salsa, [Rice, Lettuce, Guacamole...       11.25  
960   [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...        8.75  
1316  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...        8.75  
1884  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
2156  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.25  
2223      [Roasted Chili Corn Salsa, Fajita Vegetables]        8.75  
2269  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...        8.75  
2683  [Roasted Chili Corn Salsa, [Fajita Vegetables,...        8.75  
3223  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.25  
3293  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...        8.75  
4109  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.25  
4201  [Fresh Tomato Salsa, [Fajita Vegetables, Black...       11.25  
4261  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
4541  [Tomatillo Green Chili Salsa, [Fajita Vegetabl...        8.75  
4573  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...        8.75

### Paso 9. ¿Cuántas veces pidió alguien más de un refresco en lata?

1º filtramos el df donde el nombre es igual a Canned Soda

In [48]:
filtro_2=chipo[chipo['item_name']=='Canned Soda']
filtro_2

order_id  quantity    item_name choice_description  item_price
18           9         2  Canned Soda           [Sprite]        2.18
28          14         1  Canned Soda       [Dr. Pepper]        1.09
51          23         2  Canned Soda     [Mountain Dew]        2.18
53          24         1  Canned Soda           [Sprite]        1.09
107         47         1  Canned Soda       [Dr. Pepper]        1.09
...        ...       ...          ...                ...         ...
3897      1562         1  Canned Soda     [Mountain Dew]        1.09
3926      1575         1  Canned Soda       [Dr. Pepper]        1.09
3936      1578         1  Canned Soda  [Diet Dr. Pepper]        1.09
4008      1604         1  Canned Soda        [Diet Coke]        1.09
4051      1621         1  Canned Soda           [Sprite]        1.09

[104 rows x 5 columns]

2º contar las ocurrencia por order_id 

In [51]:
pedido=filtro_2.groupby('order_id').size()
pedido

order_id
9       1
14      1
23      1
24      1
47      1
       ..
1562    1
1575    1
1578    1
1604    1
1621    1
Length: 94, dtype: int64

2º filtrar dodne `Canned Soda` aparezca en mas de una vez en la columna order_id

In [52]:
total_pedido=pedido[pedido>1]
total_pedido

order_id
81      2
108     3
450     2
496     2
787     2
1156    2
1396    2
1434    2
1484    2
dtype: int64

In [53]:
len(total_pedido)

9

 9 personas pidieron mas de una vez Canned Soda